text/0000-user-ids.md

```diff
@@ -0,0 +1,42 @@
+# Secure runtime environments
+
+## Summary
+
+This RFC recommends that Paketo Buildpacks should in general strive to avoid making run-time modifications to the Buildpacks layers or the application directory by default. Instead, they should consider the output image to be read-only by default and explicitly set the parts that need to be writeable during runtime.
+
+
+## Motivation
+
+As of Buildpacks [RFC 0085](https://github.com/buildpacks/rfcs/blob/main/text/0085-run-uid.md) the Cloud Native Buildpacks project recommends that the run-time and build-time user ids be different. The motivation behind this RFC was to improve the security of output images and prevent arbitrary runtime modifications to built artifacts.
+
+The run-time and build-time user can potentially share the same group id if they wish to make modifications to the output image during run-time, by changing the permissions of the files they wish to modify during run-time to be group writable during the build process. Ideally most buildpacks should use the temporary directories or other known scratch directories for ephemeral runtime files.
```

Comment:
>Ideally most buildpacks should use the temporary directories

What worries me about relying on temp directories is that these can be small and limited in capability. It's not uncommon for `/tmp` to be mounted with `noexec`, `nosuid`, and `nodev`, with `noexec` being the most likely problem (i.e. you can run something from there).

I don't really know if this is common in K8s though. Is that something you or other's have seen? Perhaps this is something we should investigate, as this would impact how buildpacks are configured to work.

Reply:
Sorry, I missed this comment. `noexec` means you can't execute a script from that location. I'm not sure if this is a scenario that comes up, but if you needed to write out a script to run and you can only write to `/tmp`, then you could not execute the script because of `noexec`.

I don't think Java buildpacks are doing this, and it's probably not a concern, but thought I would point it out and see if anyone else things that might affect their buildpacks. If not, feel free to resolve this.

Reply:
I'm not thinking of a good reason why you'd want to do this. I think if you had an app that was going to shell out and run some script, it would likely contain that script in the app. But let's say that you had an app that generated some commands to run as a script, wrote it out, and run the script. That would not work. Probably not a big deal, so I'm OK closing out this comment thread.

>buildpacks should be able to accomplish anything they need to with exec.d right?

IMHO No. You can't modify start commands from exec.d. It doesn't come up often, but I know that has been limiting at least once. Not worth discussing here. Just me ranting a little.